### PR TITLE
Cosmos noise realisation

### PR DESCRIPTION
**Pull request by b-remy**
This PR aims to solve #43 and #44.
After this PR, it will be possible to sample simulations with cosmos noise realisation or gaussian noise realisation.

### CODE REVIEW

File: papers/Remy2021/sample_cosmos_jz.job [hunk 1]
```diff
@@ -24,9 +25,8 @@ module load tensorflow-gpu/py3/2.5.0
 
 for i in {1..6}
 do
-  srun python ./sample_ODE.py --convergence=../data/ktng/ktng_kappa360v2.fits\
+  srun python ./sample_maps.py --convergence=../data/ktng/ktng_kappa360v2.fits\
                         --mask=../data/COSMOS/cosmos_full_mask_0.29arcmin360copy.fits\
-                        --num_steps=10\
                         --sigma_gamma=0.148\
```
Comment (EiffL) [inline papers/Remy2021/sample_cosmos_jz.job:30]: what do we need sigma_gamma for?

File: scripts/sample_maps.py [hunk 2]
```diff
@@ -62,22 +73,23 @@ def log_gaussian_prior_b(map_data, sigma):
...
 
+  def log_likelihood(x, sigma, meas_shear, sigma_mask):
+    """ Likelihood function at the level of the measured shear
+    """
+    x = x.reshape((360, 360,2))
+    ke = x[...,0]
+    kb = x[...,1]
+    model_shear = jnp.stack(ks93inv(ke, kb), axis=-1)
+  
+    return - jnp.sum((model_shear - meas_shear)**2/((sigma_gamma)**2 + sigma**2 + sigma_mask) )/2.
```
Comment (EiffL) [inline scripts/sample_maps.py:90]: Does std1 already contain a high variance of masked regions?

File: scripts/sample_maps.py [hunk 3]
```diff
@@ -133,16 +145,19 @@ def main(_):
...
     # Add noise the shear map
-    gamma1 += FLAGS.sigma_gamma * onp.random.randn(map_size,map_size)
-    gamma2 += FLAGS.sigma_gamma * onp.random.randn(map_size,map_size)
+    if FLAGS.cosmos_noise_realisation:
+      print('cosmos noise real')
+      gamma1 += fits.getdata(FLAGS.cosmos_noise_e1).astype('float32')
+      gamma2 += fits.getdata(FLAGS.cosmos_noise_e2).astype('float32')
+
+    else:
+      gaussian_noise = fits.getdata(FLAGS.gaussian_noise).astype('float32')
```
Comment (EiffL) [inline scripts/sample_maps.py:154]: hummmm.... ok I see.... ytou could also have set the seed with a the jax.random.normal right?

Comment (EiffL) [reply scripts/sample_maps.py:154]: but this should work :-)

**Review by EiffL — COMMENTED** [review]
just a few questions

File: scripts/sample_maps.py (hunk 2 above)
Comment (b-remy) [reply scripts/sample_maps.py:90]: `std1` contains zeros in masked regions. High variance is constained in `sigma_mask`

File: papers/Remy2021/sample_cosmos_jz.job (hunk 1 above)
Comment (b-remy) [reply papers/Remy2021/sample_cosmos_jz.job:30]: it is useless with the ODE sampler indeed

File: scripts/sample_maps.py (hunk 3 above)
Comment (b-remy) [reply scripts/sample_maps.py]: I agree that using `jax.random.normal` with constant seed seems the best way to do this. It's fixed

**Comment by b-remy** [comment]
If @EiffL agrees, this PR can be closed :-)
